lib/codenav.py

```diff
@@ -72,16 +73,40 @@ class RelatedCodeLinePhantom:
     _template = None
 
     def __init__(self):
+        self._lock = Lock()
```

Comment:
Could you document exactly which variables are protected by this lock, and perhaps colocate the lock definition with the definition of those variables?

Just calling this `_lock` is vague.

Reply:
Well, it mainly protects `_decorate_locked()` and most of `on_modified()`, but since these method can mutate any of the member variables, `_lock` protects everything under it. I decided to lock at this level since it was easiest to reason about, though I'm not sure it's the best or clearest approach. Do you think something else should be done, or should I leave a comment above `_lock` explaining this?

Reply:
So if it protects all member variables, then you must ensure that *any* access to the member variables are protected by the lock (not just in those two methods). I haven't checked this (you may already be doing this), but fyi.

Python does guarantee that statements like `self.foo = bar` are atomic, regardless of locking, so those types of operations need not acquire the lock.

So:
- Formalize exactly which variables should be protected by the lock (possibly all of them)
- *Every* access or modification of that variable should first acquire the lock, unless it's simply an atomic assignment.
    * It might be clearer to acquire the lock even for atomic assignments, since then the reader doesn't need to have a very deep understanding of Python semantics.